src/hotspot/cpu/ppc/macroAssembler_ppc.cpp

```diff
@@ -2832,15 +2842,46 @@ void MacroAssembler::compiler_fast_unlock_object(ConditionRegister flag, Registe
   }
 #endif
 
-  ld(displaced_header, in_bytes(ObjectMonitor::recursions_offset()), current_header);
+  if (LockingMode == LM_LIGHTWEIGHT) {
+    // Some other platforms use a C2HandleAnonOMOwnerStub stub.
+    // This doesn't work well here, because we reach here for native wrappers, too.
+    // In addition, the stub may be out of range for conditional branches.
+    // The fixup code is not long, so we do it among other special cases below.
+
+    // If the owner is anonymous, we need to fix it.
+    andi_(R0, temp, ObjectMonitor::ANONYMOUS_OWNER);
+    bne(CCR0, anonymous);
+  }
```

Comment:
Let's see if @rkennke remembers. Roman I wonder why you've implemented the C2HandleAnonOMOwnerStub instead of just taking the slow path into the runtime? @TheRealMDoerr wants to implement the logic inline and I'm actually worried if that will pay off the cost.
Maybe we should add an illegal instruction to see if it's reached. Do you know a benchmark that would be good for such a test?

Reply:
> Let's see if @rkennke remembers. Roman I wonder why you've implemented the C2HandleAnonOMOwnerStub instead of just taking the slow path into the runtime? @TheRealMDoerr wants to implement the logic inline and I'm actually worried if that will pay off the cost. Maybe we should add an illegal instruction to see if it's reached. Do you know a benchmark that would be good for such a test?

Handling the anonymous owner there is a rather rare situation. Handling this inline is possible, but it goes against static branch prediction: branch predictors in modern CPUs (at least in x86 and aarch64, not sure about ppc) assume that forward-branches are commonly *not* taken. However, we would very often branch forward in this situation and tend to cause a pipeline stall. It is better to leave the common path clean and short, and branch out to the stub in the uncommon case of handling the anonymous owner. Some of the renaissance benchmarks have been rather sensitive to the performance of monitor-locking (but I don't remember which ones ;-) ).

Reply:
Sorry I was unclear about what I meant with 'inline': I meant part of the instruction sequence that `FastUnlock` is expanded to which isn't just a basic block. In that sequence it is out-of-line because it is at the end. So the forward branch that needs to be taken if it has an anonymous owner will be well predicted: not taken; continue with the next instruction.
Reading again the inflation coding I see that there was a misunderstanding on my side. I though the actual owner will be set right when the monitor was successfully installed, but this isn't done of course, because finding the actual owner is expensive with LM_LIGHTWEIGHT locking. Sorry my bad.
So it is not as unlikely as I thought, I'm still not totally convinced though that it is beneficial to handle the anonymous owner as part of `FastUnlock` case but I won't object against it.

Reply:
> Sorry I was unclear about what I meant with 'inline': I meant part of the instruction sequence that `FastUnlock` is expanded to which isn't just a basic block. In that sequence it is out-of-line because it is at the end. So the forward branch that needs to be taken if it has an anonymous owner will be well predicted: not taken; continue with the next instruction. Reading again the inflation coding I see that there was a misunderstanding on my side. I though the actual owner will be set right when the monitor was successfully installed, but this isn't done of course, because finding the actual owner is expensive with LM_LIGHTWEIGHT locking. Sorry my bad. So it is not as unlikely as I thought, I'm still not totally convinced though that it is beneficial to handle the anonymous owner as part of `FastUnlock` case but I won't object against it.

Ah I understand what you mean. Handling the anonymous owner in FastUnlock is necessary for correctness. If we unlock a monitor and observe that it is locked anonymously, we must 1. update the owner and 2. pop the object from the lock-stack. The alternative here would be to detect this case and branch into the runtime to handle this, but handling it in compiled code is much more lightweight and not too difficult to do, nor does it add all that much to the code-size. Infact, this is very likely smaller, because calling to the runtime needs a lot of scaffolding to push/pop all registers, etc.

Reply:
Thanks for your explanation!

Reply:
> The alternative here would be to detect this case and branch into the runtime to handle this, but handling it in compiled code is much more lightweight and not too difficult to do, nor does it add all that much to the code-size. Infact, this is very likely smaller because calling to the runtime needs a lot of scaffolding to push/pop all registers, etc.

But Roman of course the RT call wouldn't be part of the FastUnlock. Instead FastUnlock would return NE to indicate failure and then the slow path to the RT would be taken. So this is the smaller version.
You'll have lots of uncontended unlocks with dead code. And you'll handle the anonymous case just once / inflation on the other hand.
But everything's said now. I'm ok with it.

Reply:
Ok, you are right! The anon case would just branch to the exit (zero flag would readily be un-set by the test, right?). The normal case would fall-through. This sounds like a useful simplification, and should not affect performance. I'll probably do that in x86_64 and aarch64, feel free to also simplify the PPC version.

Reply:
Correct that's what I meant. I was just not sure if there is a reason not to do that but have a dedicated C2HandleAnonOMOwnerStub to handle that case.

Reply:
I think we don't even need to check the `ANONYMOUS_OWNER` bit because that case is also handled by
```
  cmpd(flag, temp, R16_thread);
  bne(flag, failure);
```
If `temp` has the `ANONYMOUS_OWNER` set it's unequal to the current thread and we jump to failure with CR0 NE.

Reply:
You are right.

Reply:
I have removed it and will rerun tests. Please take a look. And thanks for the discussion.

Reply:
I had just a quick look. Looks good.